app/models/cost_query/cache.rb

```diff
@@ -0,0 +1,75 @@
+#-- copyright
+# OpenProject Reporting Plugin
+#
+# Copyright (C) 2010 - 2014 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#++
+
+module CostQuery::Cache
+  class << self
+
+    def check
+      reset! if reset_required?
+    end
+
+    def reset!
+      CostQuery::Filter.reset!
+      CostQuery::Filter::CustomFieldEntries.reset!
+      CostQuery::GroupBy.reset!
+      CostQuery::GroupBy::CustomFieldEntries.reset!
+
+      update_reset_on
+    end
+
+    protected
+
+    attr_accessor :custom_fields_updated_on,
+                  :custom_fields_id_sum
+
+    def invalid?
+      updated_on = fetch_custom_field_updated_at
+      id_sum = fetch_custom_fields_changed
+
+      custom_fields_updated_on != updated_on ||
+        custom_fields_id_sum != id_sum
+    end
+
+    def update_reset_on
+      return if caching_disabled?
+
+      self.custom_fields_updated_on = fetch_custom_field_updated_at
+      self.custom_fields_id_sum = fetch_custom_fields_changed
+    end
+
+    def fetch_custom_field_updated_at
+      WorkPackageCustomField.maximum(:updated_at)
+    end
+
+    def fetch_custom_fields_changed
+      WorkPackageCustomField.sum(:id) + WorkPackageCustomField.count
```

Comment:
> generally [I] oppose the "this is bad because I didn't write it" movement. I consider code which ran for a long time to be generally better than code I just wrote.

Don't get me wrong here. I don't think it is bad because I did not write it. I questioned the quality here, because I did not *understand* it. Then I went along trying to understand it. and I came to the conclusion, that the algorithm used is flawed.

We do not need to discuss how likely it is that errors can occur, but either you can tell me that I did still not understand it and that this code works perfectly fine or we are (again) at the point where I have to ask for the benefit of this caching business. We have a bunch of errors we can make implementing this cache. If we **do** make errors, bugs will happen and they will be hard to reproduce. We have something like three choices:

1. Live with those bugs: They are rare, a work around is to "add a new custom field" or "restart the server", what can go wrong?
2. Fix the cache, forever: Surely, with enough time we will simply fix the errors in our cache and everyone is happy
3. Ditch the cache: You already ditched the cache for certain use-cases, if we can't prove that the cache has a noticably positive performance impact we can ditch it and don't worry anymore

I personally think we should choose Option **2** or **3**. I think we only have the time for option **3**.